src/compiler/checker.ts

```diff
@@ -8059,7 +8059,15 @@ namespace ts {
                             error(accessExpression.argumentExpression, Diagnostics.Element_implicitly_has_an_any_type_because_index_expression_is_not_of_type_number);
                         }
                         else {
-                            error(accessExpression, Diagnostics.Element_implicitly_has_an_any_type_because_type_0_has_no_index_signature, typeToString(objectType));
+                            let suggestion: string;
+                            if (propName !== undefined && (suggestion = getSuggestionForNonexistentPropertyWithName(propName as string, objectType))) {
+                                if (suggestion !== undefined) {
+                                    error(accessExpression, Diagnostics.Property_0_does_not_exist_on_type_1_Did_you_mean_2, propName as string, typeToString(objectType), suggestion);
+                                }
+                            }
+                            else {
+                                error(accessExpression, Diagnostics.Element_implicitly_has_an_any_type_because_type_0_has_no_index_signature, typeToString(objectType));
+                            }
                         }
                     }
                     return anyType;

```

Comment:
We would like to get this suggestion even without noImplicitAny on. So this same check needs to happen in the `if (indexType.flags & (TypeFlags.StringLiteral | TypeFlags.NumberLiteral))` below. In fact, `propName` being defined is equivalent to that check, so you can check that instead of the flag check, and you can use `propName as string` instead of `indexType.value`.

I think you *might* be able to share code between the branches in the following way. But I'm not sure: 
1. Change the `else` branch  on line 8061 to `else if (!propName)` so that you no longer get "Element implicitly has an any type" error for literals (and symbols), but will fall through to "Property 0 does not exist on type 1".
3. Move your new code there, replacing the existing "Property 0 does not exist on type 1" error.
4. If `accessExpression && !isConstEnumObjectType(objectType)`, issue an error only if noImplicitAny is true. The fallback error in this case should be "Element implictly has an any type" instead of "Property 0 does not exist on type 1".

I'm not sure the code will be an improvement, but it should be shorter than duplicating it, so see what you think.